openshift-tools/pkg/discover-etcd-initial-cluster/initial-cluster.go

```diff
@@ -105,173 +109,197 @@ func (o *DiscoverEtcdInitialClusterOptions) Validate() error {
+	if err != nil && dataDirExists {
+		fmt.Fprintf(os.Stderr, "failed to create etcd client, but the server is already initialized as member %q before, starting as etcd member: %v", o.TargetName, err.Error())
 		return nil
-	} else if err != nil {
-		return err
+	}
+	if err != nil {
+		return fmt.Errorf("failed to create etcd client: %v",err)
 	}
 	defer client.Close()
 
-	var targetMember *etcdserverpb.Member
-	var allMembers []*etcdserverpb.Member
+	target := url.URL{
+		Scheme: o.TargetPeerURLScheme,
+		Host:   fmt.Sprintf("%s:%s", o.TargetPeerURLHost, o.TargetPeerURLPort),
+	}
+
 	for i := 0; i < 10; i++ {
 		fmt.Fprintf(os.Stderr, "#### attempt %d\n", i)
-		targetMember, allMembers, err = o.checkForTarget(client)
 
-		for _, member := range allMembers {
-			fmt.Fprintf(os.Stderr, "      member=%v\n", stringifyMember(member))
+		// check member list on each iteration for changes
+		cluster, err := client.MemberList(context.TODO())
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "member list request failed: %v", err)
+			continue
+		}
+		logCurrentMembership(cluster.Members)
+		targetMember, found := checkTargetMember(target, cluster.Members)
+
+		// Condition: unstarted member found
+		// This member is part of the cluster but has not yet started. We know this because the name is populated at
+		// runtime which this member does not have.
+		// Result: provide initial cluster value so etcd can communicate with peers during startup.
+		if found && targetMember.Name == "" && !dataDirExists {
+			fmt.Print(getInitialCluster(o.TargetName, targetMember, cluster.Members))
+			return nil
 		}
-		fmt.Fprintf(os.Stderr, "      target=%v, err=%v\n", stringifyMember(targetMember), err)
 
-		// we're done because we found what we want.
-		if targetMember != nil && err == nil {
-			break
+		// Condition: unstarted member found with dataDir
+		// This member is part of the cluster but has not yet started yet has a dataDir
+		// Result: archive old dataDir and return Error
+		if found && targetMember.Name == "" && dataDirExists {
+			archivedDir, err := archiveDataDir(o.DataDir)
+			if err != nil {
+				return err
+			}
+			return fmt.Errorf("member %q is unstarted but previous members dataDir exists: archiving to %q", target.String(), archivedDir)
 		}
 
-		fmt.Fprintf(os.Stderr, "#### sleeping...\n")
-		time.Sleep(1 * time.Second)
-	}
+		// Condition: member found and has started in the past
+		// Result: empty initial config
+		if found && dataDirExists {
+			return nil
+		}
 
-	switch {
-	case targetMember == nil && memberDirExists:
-		// we weren't able to locate other members and need to return based previous memberDir so we can restart.  This is again the off and on flow.
-		fmt.Fprintf(os.Stderr, "Couldn't get targetMember, but memberDir %s is present on %s. Returning with no error.\n", memberDir, o.TargetName)
-		fmt.Printf(o.TargetName)
-		return nil
+		// Condition: member found and has started in the past but not has no dataDir
```

Comment:
nit: double negative can lead to confusion. 
s/not has no dataDir/has no dataDir/

Reply:
yeah typo `s/not/now` will align comments and this